serverless/azure/src/get.ts

```diff
@@ -0,0 +1,124 @@
+    super(message);
+  }
+}
+
+async function nativeDecompress(
+  buf: ArrayBuffer,
+  compression: Compression
+): Promise<ArrayBuffer> {
+  if (compression === Compression.None || compression === Compression.Unknown) {
+    return buf;
+  } else if (compression === Compression.Gzip) {
+    let stream = new Response(buf).body!;
+    let result = stream.pipeThrough(new DecompressionStream("gzip"));
+    return new Response(result).arrayBuffer();
+  } else {
+    throw Error("Compression method not supported");
+  }
+}
+
+//const CACHE = new ResolvedValueCache(25, undefined, nativeDecompress);
+const CACHE = new SharedPromiseCache(25, undefined, nativeDecompress);
+
+export async function getZxy(
+  source: Source,
+  tile: [number, number, number],
+  ext: string,
+  allowed_origin: string,
+  requestedEtag: string | null
+): Promise<HttpResponseInit> {
+  const cacheableResponse = (
+    body: ArrayBuffer | string | undefined,
+    cacheable_headers: Headers,
+    status: number
+  ) => {
+    let resp_headers = new Headers(cacheable_headers);
+    if (allowed_origin)
+      resp_headers.set("Access-Control-Allow-Origin", allowed_origin);
+    resp_headers.set("Vary", "Origin");
+    resp_headers.set(
+      "Cache-Control",
+      "max-age=" + (process.env["CACHE_MAX_AGE"] ?? 86400)
+    );
+    return { body, headers: resp_headers, status: status };
+  };
+
+  const cacheableHeaders = new Headers();
+
+  const p = new PMTiles(source, CACHE, nativeDecompress);
+  try {
+    const p_header = await p.getHeader();
+
+    if (requestedEtag && requestedEtag === p_header.etag) {
+      return cacheableResponse(undefined, cacheableHeaders, 304);
+    }
+
+    if (tile[0] < p_header.minZoom || tile[0] > p_header.maxZoom) {
+      return cacheableResponse(undefined, cacheableHeaders, 404);
+    }
+
+    for (const pair of [
+      [TileType.Mvt, "mvt"],
+      [TileType.Png, "png"],
+      [TileType.Jpeg, "jpg"],
+      [TileType.Webp, "webp"],
+    ]) {
+      if (p_header.tileType === pair[0] && ext !== pair[1]) {
+        if (p_header.tileType == TileType.Mvt && ext === "pbf") {
+          // allow this for now. Eventually we will delete this in favor of .mvt
+          continue;
+        }
+        return cacheableResponse(
+          `Bad request: requested .${ext} but archive has type .${pair[1]}`,
+          cacheableHeaders,
+          400
+        );
+      }
+    }
+
+    const tileData = await p.getZxy(tile[0], tile[1], tile[2]);
+
+    switch (p_header.tileType) {
+      case TileType.Mvt:
+        cacheableHeaders.set("Content-Type", "application/x-protobuf");
+        break;
+      case TileType.Png:
+        cacheableHeaders.set("Content-Type", "image/png");
+        break;
+      case TileType.Jpeg:
+        cacheableHeaders.set("Content-Type", "image/jpeg");
+        break;
+      case TileType.Webp:
+        cacheableHeaders.set("Content-Type", "image/webp");
+        break;
+    }
+
+    if (p_header.etag) {
```

Comment:
I don't think we want this. If I am reading this correctly it will pass the etag of the entire archive as the etag of every tile; this means the etag is no longer properly content-aware since tiles will almost always be different.

Reply:
No, I've changed that to use the tileData.etag, but haven't pushed the change. I've also added a `Last-Modified`, which is the reason I haven't pushed it yet. I need #304 to be merged first. 

But I don't think this is a problem, because the source pmtiles file will not change, so every tile can have the same ETag. It actually will have the same ETag, at least if you use `FetchSource` or `AzureStorageSource` because the ETag comes from the file in storage, not the bytes. The ETag is unique in combination with the URL, so it will work in browsers. If you change pmtiles file, the ETag will change for all tiles.